strapi/strapi
# Fix documentation plugin overrides

### What does it do?

Fixes the overrides
- Migrates overrides to its own service
- Adds excludeFromGeneration function 
- Adds a mutateDocumentation function that can be passed to the documentation plugin's config
- Fixes overrides running no matter what (via specifying pluginOrigin, or using excludeFromGeneration)
- Fixes upload plugin override and registers it from the plugin
- Adds ability to only run overrides for specific version
- Adds tests for the overrides

### Why is it needed?

To resolve bugs and ensure the overrides work as expected

### How to test it?
Run yarn setup

Make sure the tests pass with
```
yarn nx run @strapi/plugin-documentation:test:unit
```
Test manually:
- Overrides are already being provided for upload and users-permissions
  - Confirm the Swagger UI reflects the overrides registered in packages/core/upload/documentation/content-api.json
  - Confirm the same for users-permissions in packages/plugins/users-permissions/documentation/content-api.yaml
  - Both cases exclude generation so make sure only what is described in these files ends up in the UI
  - Remove the excludeFromGeneration in `packages/plugins/users-permissions/server/register.js` or `packages/plugins/upload/server/register.js`, and confirm there is a mix of generated docs and the overrides (there should be additional schema components)
- Test overrides don't run for plugins not listed in the config
  - In `examples/getstarted/config/plugins.js` try each of the following
```
// config/plugins.js
module.exports = {
	documentation: {
		enabled: true,
		config: {
		  "x-strapi-config": {
			// Default
		    "plugins": [
		      "upload",
		      "users-permissions"
		    ],
			// Custom
		    "plugins": [
		      "upload",
		    ],
			// Do not generate for plugins
		    "plugins": []
		  }
		}
	}
}
```
- Test the override only runs for a specific version
	- Add the following to an override and then change the config to a different version
```
// upload override content-api.json
// Only run this override for version 1.0.0
info: { version: '1.0.0' } 

// config/plugins.js
module.exports = {
	config: {
		info: {version: '2.0.0'}
	}
}
```
- Test you can mutate the document directly after generation and overrides
```
// config/plugins.js
module.exports = {
	documentation: {
		config: {
			'x-strapi-config': {
				mutateDocumentation: (generatedDocumentationDraft) => {
        			generatedDocumentationDraft.paths['/answer-to-everything'].get.responses['200'].description = "42";
      			},
			}		
		}
	}
}
```


### Related issue(s)/PR(s)

https://github.com/strapi/strapi/issues/12522
https://github.com/strapi/strapi/issues/8629
